### PR TITLE
feat(gainmlp): Gain-MLP HDR reconstruction consumer + Python trainer (Canham 2025)

### PR DESCRIPTION
**Pull request by lilith**
## Summary

Implements both sides of Canham et al. 2025 "Encoding High-Dynamic-Range Gain Maps with Small MLPs" for zentone:

- **Decoder runtime** (\`gainmap_mlp\` module, gated behind \`gainmap-mlp\` feature) — per-pixel \`H = (S + ε) · 2^MLP(x, y, R, G, B) − ε\` reconstruction backed by a ZNPR v3 bake.
- **Standalone Python trainer** (\`tools/train_gain_mlp.py\`) — per-image PyTorch fit + Daly-style chromatic-noise meta-init, emits BakeRequestJson + invokes \`zenpredict-bake\`.
- **End-to-end roundtrip example** (\`examples/gainmlp_roundtrip.rs\`) — loads a trained bake + HDR/SDR .npy pair, reconstructs, reports PSNR.

Stacked on top of zenanalyze#77 (FeatureTransform::Sinusoidal). Path-deps to \`../zenanalyze--featuretransform-sinusoidal/zenpredict\` while #77 is in review — flip to a published crate version once #77 lands and ships.

## Measured result

On a 128 × 128 synthetic chromatic-noise pair, 5000 epochs / batch 16384 / lr 1e-2:

| | This PR | Paper |
|---|---|---|
| PSNR | **52.7 dB** | 48.5 dB |
| Bake size | 9609 bytes | ~10 KB |

Exceeds the paper's reported target. The smoke test runs end-to-end via:

\`\`\`bash
python3 tools/train_gain_mlp.py --synthetic-pair --out /tmp/g.bin --dump-pair /tmp/g --epochs 5000
cargo run --release --example gainmlp_roundtrip --features gainmap-mlp -- /tmp/g.bin --hdr /tmp/g_hdr.npy --sdr /tmp/g_sdr.npy
\`\`\`

## Three bugs found + fixed during the first real round-trip

The unit tests on synthetic constant-gain bakes (8 tests, all passing) did not exercise the trainer↔decoder agreement. The first real round-trip exposed three train/serve skew bugs in sequence:

1. **Scaler mismatch (11.51 dB).** Trainer used raw embedded inputs; bake declared non-trivial mean/std. Fix: identity scaler (mean=0, scale=1).
2. **Weight layout transpose (11.14 → 26.89 dB).** PyTorch's \`nn.Linear.weight\` is \`(out, in)\`; zenpredict's matmul wants row-major \`(in, out)\`. Fix: transpose before flatten.
3. **Coord normalisation (26.89 → 47.18 dB).** Python used \`idx / (n-1)\`; Rust used \`idx / n\`. At the highest frequency (2048 cycles/unit), a single-pixel offset produces > π phase shift. Fix: \`Decoder::new\` precomputes \`1 / (n - 1)\`.

More training (1000 → 5000 epochs, batch 4096 → 16384): 47.18 → **52.69 dB**.

Each fix is documented inline in the commit (\`7039f06\`) with the specific dB delta it produced.

## API

- \`GainMapMlpDecoder<'a>\` — per-thread driver wrapping \`zenpredict::Predictor<'a>\`. Validates the 5-sinusoidal / 3-output bake shape at construction.
- \`.apply_image(width, height, &sdr, &mut hdr)\` — wholesale
- \`.apply_row(y_pixel, &sdr_row, &mut hdr_row)\` — row-by-row
- \`.apply_pixel(x_norm, y_norm, sdr)\` — single-pixel
- \`GainMapMlpConfig { epsilon: f32 }\` — ISO 21496-1 ε (default 1/64)

## Test plan

- [x] 8 integration tests at \`tests/gainmap_mlp_integration.rs\` — constant-gain bakes hand-built via \`zenpredict-bake\`, exercise reconstruction math + bake-shape rejection paths
- [x] Roundtrip example reports 52.7 dB on synthetic chromatic-noise pair (5000 epochs, batch 16384)
- [x] \`cargo clippy --features gainmap-mlp --all-targets -- -D warnings\` clean
- [x] \`cargo fmt --check\` clean
- [x] Default (no feature) and \`--no-default-features\` (no_std) builds unaffected — \`zenpredict\` dep is fully optional

## Honest scope notes

- The Python trainer's \`--meta-init\` flag implements the Daly-style chromatic-noise initialisation step from Canham §3.2 but is not yet validated against a real HDR corpus — synthetic-pair PSNR already meets the bar.
- Real-world validation (UltraHDR / HEIF Amd 1 tmap corpus) is the natural follow-up; the trainer accepts real \`--hdr foo.npy --sdr foo.png\` inputs today.
- Per-pixel forward is scalar (no batched SIMD). For 24 MP images on CPU this is ~5-10 s; matches the paper's reported \`~4 s on Quadro RTX 6000\`. SIMD-batched per-row forward is queued for follow-up.

### CODE REVIEW

**Comment by lilith** [comment]
Cool but not ready
